### ISSUE TITLE
Should we continue advocating the use of algolia.ttAdapter?

### ISSUE THREAD

**Issue by vvo** [algolia/algoliasearch-client-js#106]
We currently have a function called ttAdapter in the JavaScript client.

A recent issue (#105) about it came . To me, it brings another issue which is: should we keep [advocating](https://www.algolia.com/doc/tutorials/auto-complete#TutorialUI) the use of our ttAdapter function? 

We are getting issues about typeahead in the algolia search JavaScript client source code.

Maybe we could just tell in our tutorials: "If you want to use algolia with typeahead, use this code:"

```js
function ttSource(query, cb) {
  self.search(query, params, function(err, content) {
    if (err) {
      cb(err);
      return;
    }

     cb(content.hits);
  });
};
```

Btw, if you are using typeahead v2...., use this:..

Having a shortcut it in the JS client only helps our tutorials to be more succinct but nothing amazing is done as we can see in https://github.com/algolia/algoliasearch-client-js/blob/f279b21da90be745a34e2c29bfec346897583998/src/AlgoliaSearch.js#L1351-L1363

Having a dependency on typeahead in the js client is a historical decision but I guess we should hide it now and ask people to put the dependency in their code. And explain to them how to plug typeahead with 5 lines of code instead of giving them a small function.

**Comment by vvo** [comment]
thoughts @redox @bobylito @pixelastic?

**Comment by kokliKo** [comment]
I think it's a good idea. A lot of customers don't understand what ttAdapter is.

**Comment by redox** [comment]
I agree; this code should not be in the JS client and was indeed an historical decision. Some documentation is probably enough for the `< 0.11` versions but it seems more complex since `0.11` -> maybe we should build a library `algoliasearch-typeahead.js` or something like that hiding the complexity.

**Comment by pixelastic** [comment]
I agree, the additional step to make it work with typeahead is minimal, so this could fit in a tutorial easily. We could mark this method as deprecated for a few versions (`console.log` that it's deprecated) and then remove it completly (and bumping the semver accordingly).

Not sure of the benefits of an `algoliasearch-typehead` build, though. The difference between the two will be very small.

**Comment by vvo** [comment]
@pixelastic we will try not to bump major to 4 because it would require us to do a lot of communication and then we would have to support both 2.x 3.x 4.x. So we are going to deprecate it and remove all references to ttAdapter in our docs.

**Comment by redox** [comment]
I would wait until we have a plan to move to v4 before introducing the `console.log` + deprecations, otherwise it will be super floody.

TL;DR:
 - `ttAdapter` will be removed from v4
 - we cannot break the backward compat in v3
 - we should then maybe support typeahead `0.11`, otherwise it will generate a lot of questions

@pixelastic about the `algoliasearch-typeahead`, I was not speaking about a build of the JS client including it; I was more speaking about a small helper exposing a `ttAdapter` method. Are we on the same page?

**Comment by pixelastic** [comment]
You're right, we're already on v3, let's not bump the major so easily.

I'm for supporting typeahead `v0.11+` in our `v3`, and then move all typehead-related methods into a helper when we'll bump to v4.

@redox I thought it was a custom build, my bad. I'm ok with a simple helper file.

**Comment by bobylito** [comment]
:+1: to remove it. 
:+1: not in this major version.
:+1: To remove any reference to it in our docs even though we keep the code for the v3

**Comment by rhokstar** [comment]
As long as there's some tutorial on how to make algolia work with typeahead and future versions, I'm all for it.

That's how I found algolia and just this last Friday, I decided to subscribe to the starter package. Tutorials are key to adoption from a marketing point of view. Your company just got $18.3 million in Series A funding. Making it easy to adopt is the best thing to do. Keep up the good work.

**Comment by vvo** [comment]
So we will update our docs etc once typeahead 0.11 is fixed, right now it's considered as beta.

Especially twitter/typeahead.js#1224 is really annoying and breaks most implementations.

@rhokstar we released a new algoliasearch version (3.5.0) which fixes compatibility with typeahead 0.11 (but I would stick to 0.10)

```sh
npm install algoliasearch@3.5.0 --save
```

**Comment by rhokstar** [comment]
@vvo Ok, I will continue to use 0.10 because I need to get my experiment out to my customers and get usage data. But for future implementations I am going to use 0.11 which I'm going to be developing soon with some new features. Thanks for being responsive, this is good customer service!
